src/network/event_loop.rs

```diff
@@ -0,0 +1,268 @@
+                if let Some(cause) = cause {
+                    match cause {
+                        // remove peers with failed connections
+                        ConnectionError::IO(_) | ConnectionError::Handler(_) => {
+                            self.swarm.behaviour_mut().kademlia.remove_peer(&peer_id);
+                        } // ignore Keep Alive timeout errors
+                        // allow redials for this type of error
+                        _ => {}
+                    }
+                }
+            }
+            SwarmEvent::OutgoingConnectionError { peer_id, .. } => {
+                // which ever error that was,
+                // all the currently implemented ones are pretty critical
+                // remove error producing peer from further dialing
+                if let Some(peer_id) = peer_id {
+                    trace!("Error produced by peer with PeerId: {peer_id:?}");
+                    self.swarm.behaviour_mut().kademlia.remove_peer(&peer_id);
+                }
+            }
+            SwarmEvent::ConnectionEstablished { endpoint, .. } => {
+                // in case that we're listener,
+                // wait for a first successful connection
+                if endpoint.is_listener() {
+                    // check if there is a command waiting for a response
+                    let local_peer_id = self.swarm.local_peer_id();
+                    if let Some(SwarmChannel::ConnectionEstablished(ch)) =
+                        self.pending_swarm_events.remove(local_peer_id)
+                    {
+                        // signal back that we have successfully established a connection
+                        _ = ch.send(());
+                    }
+                }
+            }
+            SwarmEvent::NewListenAddr { address, .. } => {
+                let local_peer_id = *self.swarm.local_peer_id();
+                info!(
+                    "Local node is listening on: {:?}",
+                    address.with(Protocol::P2p(local_peer_id))
+                )
+            }
+            _ => {}
+        }
+    }
+
+    async fn handle_command(&mut self, command: Command) {
+        match command {
+            Command::StartListening {
+                addr,
+                response_sender,
+            } => {
+                _ = match self.swarm.listen_on(addr) {
+                    Ok(_) => response_sender.send(Ok(())),
+                    Err(err) => response_sender.send(Err(err.into())),
+                }
+            }
+            Command::Bootstrap { response_sender } => {
+                match self.swarm.behaviour_mut().kademlia.bootstrap() {
+                    Ok(query_id) => {
+                        self.pending_kad_queries
+                            .insert(query_id, QueryChannel::Bootstrap(response_sender));
+                    }
+                    // no available peers for bootstrap
```

Comment:
Do we still need to deploy 2 clients initially?

Reply:
Not necessarily. Upon deployment, Bootstraps will wait for someone to dial them, and they will commence bootstraping after a successful connection.